### PR TITLE
Update handler and tests

### PR DESCRIPTION
**Pull request by matas-cc**
### Description

Added an update endpoint to encryptonize. Nothing really innovative, mostly reusage of Store and Retrieve functionality. 

### Parent Issue
Issue #186 

### Related Pull Requests

https://github.com/cyber-crypt-com/documentation/pull/100

### Comments for the Reviewers

Some e2e tests yet to be added.

### Type(s) of Change (Split the PR if you check many)
- [ ] Added user feature
- [x] Added technical feature
- [x] Added tests
- [ ] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Infrastructure changes (Terraform, GCP, K8s, other)
- [ ] CI/CD workflow changes

### Testing Checklist
- [ ] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [x] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make docker-up; make tests`.
- [ ] I have updated relevant workflows and deployment tools.
- [x] I have updated/added relevant documentation (readme, doc comments, etc).
- [ ] I have added the license to new source code files.
- [ ] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: encryption-service/interfaces/interfaces.go [hunk 1]
```diff
@@ -78,6 +78,9 @@ type CryptorInterface interface {
 	// Encrypt encrypts data + aad with a random key and return the wrapped key and the ciphertext
 	Encrypt(data, aad []byte) (wrappedKey, ciphertext []byte, err error)
 
+	// Encrypt encrypts data + aad with a wrapped key and returns the ciphertext
+	EncryptWithKey(data, aad, key []byte) (ciphertext []byte, err error)
```
Comment (Retler) [inline encryption-service/interfaces/interfaces.go:82]: Does this method need to be public and a part of the interface? If it is only used internally for by the "Update" function, then I think it's best to remove it from the interface and make it private

File: encryption-service/tests/grpce2e/client.go [hunk 2]
```diff
@@ -110,6 +110,22 @@ func (c *Client) Retrieve(oid string) (*enc.RetrieveResponse, error) {
 	return retrieveResponse, nil
 }
 
+func (c *Client) Update(oid string, plaintext, associatedData []byte) (*enc.UpdateResponse, error) {
```
Comment (Retler) [inline encryption-service/tests/grpce2e/client.go:113]: I'm not sure if this is the case, but is the access object somehow affected by the update? I.e. can updating the object change something about the access rights?

File: encryption-service/interfaces/interfaces.go (hunk 1 above)
Comment (matas-cc) [reply encryption-service/interfaces/interfaces.go:82]: We rely on the CryptorInterface in enc package, so even though it is only used in Update, has to be part of of the interface and be Public/Exported

File: encryption-service/tests/grpce2e/client.go (hunk 2 above)
Comment (matas-cc) [reply encryption-service/tests/grpce2e/client.go:113]: I don't think so, but we could consider a test for that, I'll add it

**Review by Retler — APPROVED** [review]
Nice with the extra test for checking that the permissions are intact after update

File: encryption-service/interfaces/interfaces.go [hunk 3]
```diff
@@ -78,6 +78,9 @@ type CryptorInterface interface {
 	// Encrypt encrypts data + aad with a random key and return the wrapped key and the ciphertext
 	Encrypt(data, aad []byte) (wrappedKey, ciphertext []byte, err error)
 
+	// Encrypt encrypts data + aad with a wrapped key and returns the ciphertext
```
Comment (philip-cc) [inline encryption-service/interfaces/interfaces.go:81]:
```suggestion
	// EncryptWithKey encrypts data + aad with a wrapped key and returns the ciphertext
```

File: encryption-service/services/authn/token_middleware_test.go [hunk 4]
```diff
@@ -60,7 +60,7 @@ func CreateUserForTests(c interfaces.CryptorInterface, userID uuid.UUID, scopes
 // It ONLY tests the middleware and assumes that the LoginUser works as intended
 func TestCheckAccessTokenGoodPath(t *testing.T) {
 	userID := uuid.Must(uuid.NewV4())
-	userScope := users.ScopeRead | users.ScopeCreate | users.ScopeIndex | users.ScopeObjectPermissions
+	userScope := users.ScopeRead | users.ScopeCreate | users.ScopeIndex | users.ScopeObjectPermissions | users.ScopeUpdate
```
Comment (philip-cc) [inline encryption-service/services/authn/token_middleware_test.go:63]:
```suggestion
	userScope := users.ScopeRead | users.ScopeCreate | users.ScopeIndex | users.ScopeObjectPermissions
```
This change isn't actually necessary.

File: encryption-service/services/enc/enc.proto [hunk 5]
```diff
@@ -50,6 +53,15 @@ message RetrieveResponse{
   Object object = 1;
 }
 
+message UpdateRequest{
+  Object object = 1;
+  string object_id = 2;
+}
+
+message UpdateResponse{
+  string object_id = 1;
```
Comment (philip-cc) [inline encryption-service/services/enc/enc.proto:62]: I don't think this should return anything.

File: encryption-service/tests/grpce2e/update_test.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+// limitations under the License.
+
+package grpce2e
+
+import (
+	"bytes"
+	"testing"
+)
+
+// Test the we can store an object and retrieve it later
```
Comment (philip-cc) [inline encryption-service/tests/grpce2e/update_test.go:22]: Please update the comment to reflect the test (or remove it).

File: encryption-service/tests/grpce2e/update_test.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	_, err = client.AddPermission(oid, uid2)
+	failOnError("Add permission request failed", err, t)
+
+	client2, err := NewClient(endpoint, uat2, https)
+	failOnError("Could not create client", err, t)
+	defer closeClient(client2, t)
+
+	plaintext2 := []byte("foo2")
+
+	// Add another user to object permission list
```
Comment (philip-cc) [inline encryption-service/tests/grpce2e/update_test.go:109]: This comment needs to be updated

**Review by philip-cc — COMMENTED** [review]
Please also update `documentation/api/api-v2.md` and make sure to add a disclaimed about the concurrency issues.

File: encryption-service/interfaces/interfaces.go (hunk 3 above)
Comment (matas-cc) [reply encryption-service/interfaces/interfaces.go:81]: fixed

File: encryption-service/services/enc/enc.proto (hunk 5 above)
Comment (matas-cc) [reply encryption-service/services/enc/enc.proto:62]: Agreed, fixed

File: encryption-service/tests/grpce2e/update_test.go (hunk 6 above)
Comment (matas-cc) [reply encryption-service/tests/grpce2e/update_test.go:22]: Updated

File: encryption-service/tests/grpce2e/update_test.go (hunk 7 above)
Comment (matas-cc) [reply encryption-service/tests/grpce2e/update_test.go:109]: Updated

File: encryption-service/services/authn/token_middleware_test.go (hunk 4 above)
Comment (matas-cc) [reply encryption-service/services/authn/token_middleware_test.go:63]: :facepalm:  fixed

**Review by philip-cc — APPROVED** [review]
Ship it!

**Pull request merged**